File: Classes/NSAppearance+PBDarkMode.m
//
//  NSAppearance+PBDarkMode.m
//  GitX
//
//  Objective-C helpers for detecting the current appearance (dark/light mode).
//

#import "NSAppearance+PBDarkMode.h"
#import <objc/runtime.h>

NSString *const PBEffectiveAppearanceChanged = @"PBEffectiveAppearanceChanged";

// Shared between the KVO registration and the observing token, so the token can
// tell its own changes apart from any it gets registered for elsewhere.
static const void * const kAppearanceObservationContext = &kAppearanceObservationContext;

// AppKit routes appearance changes to NSApplication through its own
// -observeValueForKeyPath:ofObject:change:context: implementation and private
// contexts. Overriding that method on NSApplication in a category shadows
// AppKit's implementation and crashes on appearance change, so observe
// `effectiveAppearance` through this dedicated token object instead.
@interface PBAppearanceObserverToken : NSObject
@property (nonatomic, weak) id observer;
@end

@implementation PBAppearanceObserverToken

- (void)observeValueForKeyPath:(__unused NSString *)keyPath
					  ofObject:(__unused id)object
						change:(__unused NSDictionary<NSKeyValueChangeKey, id> *)change
					   context:(void *)context
{
	if (context != kAppearanceObservationContext) {
		// Not one of our observations; do not forward it up the responder
		// chain, NSObject's default implementation raises on unhandled KVO.
		return;
	}
	id observer = self.observer;
	if (observer) {
		[[NSNotificationCenter defaultCenter] postNotificationName:PBEffectiveAppearanceChanged
															object:observer];
	}
}

@end

@implementation NSAppearance (PBDarkMode)

- (BOOL)isDarkMode
{
	NSAppearanceName bestMatch = [self bestMatchFromAppearancesWithNames:@[NSAppearanceNameDarkAqua, NSAppearanceNameAqua]];
	return [bestMatch isEqualToString:NSAppearanceNameDarkAqua];
}

@end

@implementation NSApplication (PBDarkMode)

- (BOOL)isDarkMode
{
	return self.effectiveAppearance.isDarkMode;
}

- (void)registerObserverForAppearanceChanges:(id)observer
{
	PBAppearanceObserverToken *previous = objc_getAssociatedObject(self, &kAppearanceObservationContext);
	if (previous) {
		[self removeObserver:previous forKeyPath:@"effectiveAppearance"];
	}

	PBAppearanceObserverToken *token = [[PBAppearanceObserverToken alloc] init];
	token.observer = observer;
	[self addObserver:token
		   forKeyPath:@"effectiveAppearance"
			  options:0
			  context:(void *)kAppearanceObservationContext];

	// Retain the token so KVO keeps forwarding appearance changes to it.
	objc_setAssociatedObject(self, &kAppearanceObservationContext, token, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
}

@end

File: Classes/NSAppearance+PBDarkMode.swift
//
//  NSAppearance+PBDarkMode.swift
//  GitX
//
//  Converted from NSAppearance+PBDarkMode.m
//  Original by Etienne, 2018.
//
//  Note: NSAppearance+PBDarkMode.m is still compiled solely to provide the
//  `extern NSString *const PBEffectiveAppearanceChanged` definition for ObjC
//  callers.  The category implementations live here.
//

import AppKit

// Swift-side alias so Swift code can use the constant by name.
// The authoritative definition lives in NSAppearance+PBDarkMode.m.
let PBEffectiveAppearanceChanged: String = "PBEffectiveAppearanceChanged"

// MARK: - NSAppearance

extension NSAppearance {
    /// Returns `true` when this appearance resolves to Dark Aqua.
    @objc var isDarkMode: Bool {
        bestMatch(from: [.darkAqua, .aqua]) == .darkAqua
    }
}

// MARK: - NSApplication

extension NSApplication {
    /// Returns `true` when the application's effective appearance is Dark Aqua.
    @objc var isDarkMode: Bool {
        effectiveAppearance.isDarkMode
    }

    /// Registers `observer` for KVO on `effectiveAppearance` and re-posts
    /// `PBEffectiveAppearanceChanged` whenever it changes.
    /// The observation token is retained on `self` (i.e. NSApp) so it lives
    /// for the lifetime of the application — matching the original behaviour.
    @objc func registerObserverForAppearanceChanges(_ observer: Any) {
        let token = NSApp.observe(\.effectiveAppearance) { _, _ in
            NotificationCenter.default.post(
                name: Notification.Name(PBEffectiveAppearanceChanged),
                object: observer
            )
        }
        // Retain the token on NSApp so it lives for the lifetime of the app.
        objc_setAssociatedObject(
            self,
            &NSApplication.appearanceObservationKey,
            token,
            .OBJC_ASSOCIATION_RETAIN_NONATOMIC
        )
    }

    private static var appearanceObservationKey: UInt8 = 0
}
